Mathlib/NumberTheory/NumberField/FinitePlaces.lean

```diff
@@ -0,0 +1,242 @@
+/-
+Copyright (c) 2024 Fabrizio Barroero. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fabrizio Barroero
+-/
+import Mathlib.Data.Int.WithZero
+import Mathlib.NumberTheory.NumberField.Embeddings
+import Mathlib.RingTheory.DedekindDomain.AdicValuation
+import Mathlib.RingTheory.DedekindDomain.Factorization
+import Mathlib.RingTheory.Ideal.Norm.AbsNorm
+import Mathlib.Topology.Algebra.Valued.NormedValued
+
+/-!
+# Finite places of number fields
+This file defines finite places of a number field `K` as absolute values coming from an embedding
+into a completion of `K` associated to a non-zero prime ideal of `𝓞 K`.
+
+## Main Definitions and Results
+* `NumberField.vadic_abv`: a `v`-adic absolute value on `K`.
+* `NumberField.FinitePlace`: the type of finite places of a number field `K`.
+* `NumberField.FinitePlace.mulSupport_Finite`: the `v`-adic absolute value of a non-zero element of
+`K` is different from 1 for at most finitely many `v`.
+
+## Tags
+number field, places, finite places
+-/
+
+open Ideal IsDedekindDomain HeightOneSpectrum NumberField WithZeroMulInt
+
+namespace NumberField
+
+section absoluteValue
+
+variable {K : Type*} [Field K] [NumberField K] (v : HeightOneSpectrum (𝓞 K))
+
+/-- The norm of a maximal ideal as an element of `ℝ≥0` is `> 1`  -/
+lemma one_lt_norm : 1 < (absNorm v.asIdeal : NNReal) := by
+  norm_cast
+  by_contra! h
+  apply IsPrime.ne_top v.isPrime
+  rw [← absNorm_eq_one_iff]
+  have : 0 < absNorm v.asIdeal := by
+    rw [Nat.pos_iff_ne_zero, absNorm_ne_zero_iff]
+    have : Fintype (𝓞 K ⧸ v.asIdeal) := Ideal.fintypeQuotientOfFreeOfNeBot v.asIdeal v.ne_bot
+    exact Fintype.finite this
+  omega
+
+private lemma norm_ne_zero : (absNorm v.asIdeal : NNReal) ≠ 0 := ne_zero_of_lt (one_lt_norm v)
+
+/-- The `v`-adic absolute value on `K` defined as the norm of `v` raised to negative `v`-adic
+valuation.-/
+noncomputable def vadic_abv : AbsoluteValue K ℝ where
+  toFun := fun x ↦ toNNReal (norm_ne_zero v) (v.valuation x)
+  map_mul' := fun _ _ ↦ by simp only [_root_.map_mul, NNReal.coe_mul]
+  nonneg' := fun _ ↦ NNReal.zero_le_coe
+  eq_zero' := fun _ ↦ by simp only [NNReal.coe_eq_zero, map_eq_zero]
+  add_le' := by
+    intro x y
+    simp only
+    norm_cast
+    -- the triangular inequality is implied by the ultrametric one
+    apply le_trans _ <| max_le_add_of_nonneg (zero_le ((toNNReal (norm_ne_zero v)) (v.valuation x)))
+      (zero_le ((toNNReal (norm_ne_zero v)) (v.valuation y)))
+    have h_mono := StrictMono.monotone (toNNReal_strictMono (one_lt_norm v))
+    rw [← Monotone.map_max h_mono] --max goes inside withZeroMultIntToNNReal
+    exact h_mono (Valuation.map_add v.valuation x y)
```

Comment:
```suggestion
    have h_mono := (toNNReal_strictMono (one_lt_norm v)).monotone
    rw [← h_mono.map_max] --max goes inside withZeroMultIntToNNReal
    exact h_mono (v.valuation.map_add x y)
```
if you want to use dot notation. (It is a matter of taste and takes some getting used to.)